Client/Frontend/Widgets/MenuButton.swift

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import UIKit
+import BraveUI
+
+/// The button which displays a more icon (`•••`) and a set of badges
+class MenuButton: ToolbarButton {
+    /// A badge that can be added to the menu icon
+    struct Badge: Hashable {
+        var gradientView: () -> BraveGradientView
+        var icon: UIImage?
+        
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(icon)
+        }
+        
+        static func == (lhs: Self, rhs: Self) -> Bool {
+            return lhs.icon == rhs.icon
+        }
+        
+        static let playlist: Self = .init(
+            gradientView: { .gradient02 },
+            icon: UIImage(named: "playlist-menu-badge")?.withTintColor(.white)
+        )
+    }
+    
+    private(set) var badges: [Badge: UIView] = [:]
+    
+    func setBadges(_ badges: [Badge]) {
+        badges.forEach { [self] in
+            addBadge($0, animated: false)
+        }
+    }
+    
+    func addBadge(_ badge: Badge, animated: Bool) {
+        if badges[badge] != nil {
+            // Badge already exists
+            return
+        }
+        let view = BadgeView(badge: badge)
+        badges[badge] = view
+        addSubview(view)
+        if badges.count > 1 {
+            // TODO: All badges become background only
```

Comment:
does this need to be addressed in this PR?

Reply:
I got this from @kylehickinson original implementation. I'm not sure what it is for, but it isn't needed atm. It is possibly for future implementation requests.